### PR TITLE
feat(ids-api): Use syslumenn api to verify delegation.

### PR DESCRIPTION
**Pull request by valurefugl**
https://www.notion.so/Add-legal-representative-hard-check-when-user-chooses-delegation-848df5969a6d4c898f5ff648d386e71a?pvs=4

## What

Use syslumenn api to verify delegation.

## Why

To be sure that legal representative delegation still exists before authenticating.

## Screenshots / Gifs

Attach Screenshots / Gifs to help reviewers understand the scope of the pull request

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Formatting passes locally with my changes
- [ ] I have rebased against main before asking for a review

## Release Notes

- **New Features**
  - Enhanced delegation management capabilities to ensure accurate handling of delegations and their associated scopes.
  - Introduced a new method for verifying delegations at the provider level, improving the delegation verification process.
  - Added new configuration parameters for the SYSLUMENN API, enhancing service connectivity across different environments.

- **Bug Fixes**
  - Resolved issues related to delegation checks and improved handling of invalid records.

- **Documentation**
  - Updated API specifications with new endpoints and data structures for improved clarity.

- **Tests**
  - Expanded test coverage for delegation verification scenarios and authorization capabilities.

### CODE REVIEW

File: libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts [hunk 1]
```diff
@@ -256,6 +256,51 @@ export class DelegationsIncomingService {
...
+          delegationTypes.map((dt) =>
+            this.delegationsIndexService.removeDelegationRecord({
+              fromNationalId,
+              toNationalId: user.nationalId,
+              type: dt,
+              provider: AuthDelegationProvider.DistrictCommissionersRegistry,
+            }),
+          ),
+        )
+      }
```
Comment (saevarma) [inline libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts:298]: You should be able to simplify this similar to my other suggestion.

File: libs/auth-api-lib/src/lib/delegations/delegation-scope.service.ts [hunk 2]
```diff
@@ -253,6 +260,60 @@ export class DelegationScopeService {
...
+          this.delegationsIndexService.removeDelegationRecord({
+            fromNationalId,
+            toNationalId,
+            type: dt as AuthDelegationType,
+            provider: AuthDelegationProvider.DistrictCommissionersRegistry,
+          }),
+        ),
+      )
+      return []
+    }
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/delegations/delegation-scope.service.ts]: Absolutely, fixed. Although not removing from index if there is an error.

File: apps/services/auth/ids-api/src/app/delegations/delegations.controller.ts [hunk 3]
```diff
@@ -110,4 +113,26 @@ export class DelegationsController {
...
+  @Get('verify')
+  @ApiOkResponse({ type: DelegationVerificationResult })
+  async verify(
+    @CurrentUser() user: User,
+    @Headers('X-Query-From-National-Id')
+    fromNationalId: string,
+    @Query('delegationType')
+    delegationType: AuthDelegationType[],
+  ): Promise<DelegationVerificationResult> {
+    if (!Array.isArray(delegationType)) delegationType = [delegationType]
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply apps/services/auth/ids-api/src/app/delegations/delegations.controller.ts]: Fixed.

File: apps/services/auth/ids-api/src/app/delegations/delegations.controller.ts [hunk 4]
```diff
@@ -110,4 +113,26 @@ export class DelegationsController {
       delegationType,
     )
   }
+
+  @Scopes('@identityserver.api/authentication')
+  @Get('verify')
+  @ApiOkResponse({ type: DelegationVerificationResult })
+  async verify(
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply apps/services/auth/ids-api/src/app/delegations/delegations.controller.ts:122]: Fixed.

File: apps/services/auth/ids-api/src/app/delegations/delegations.controller.ts [hunk 5]
```diff
@@ -110,4 +113,26 @@ export class DelegationsController {
...
+    @CurrentUser() user: User,
+    @Headers('X-Query-From-National-Id')
+    fromNationalId: string,
+    @Query('delegationType')
+    delegationType: AuthDelegationType[],
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply apps/services/auth/ids-api/src/app/delegations/delegations.controller.ts]: Fixed.

File: libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts (hunk 1 above)
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts:298]: Done.

File: libs/auth-api-lib/src/lib/delegations/delegation-scope.service.ts [hunk 6]
```diff
@@ -253,6 +260,60 @@ export class DelegationScopeService {
     return apiScopes.map((s) => s.name)
   }
 
+  private async findDistrictCommissionersRegistryScopesTo(
+    toNationalId: string,
+    fromNationalId: string,
+    delegationTypes: string[],
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/delegations/delegation-scope.service.ts]: Ok, fixed.

File: apps/services/auth/personal-representative-public/infra/personal-representative-public.ts
[the remark this answers is not in the document]
Comment (valurefugl) [reply apps/services/auth/personal-representative-public/infra/personal-representative-public.ts:1]: Ah of course, removed.

File: apps/services/auth/ids-api/src/app/delegations/test/delegations-filters.spec.ts [hunk 7]
```diff
@@ -165,24 +165,23 @@ describe('DelegationsController', () => {
...
-        .get(path + `?delegationType=${AuthDelegationType.LegalRepresentative}`)
-        .set('X-Query-From-National-Id', testCase.fromLegalRepresentative[0])
+    it(`POST ${path} returns verified response`, async () => {
+      res = await server.post(path).send({
+        fromNationalId: testCase.fromLegalRepresentative[0],
+        delegationTypes: [AuthDelegationType.LegalRepresentative],
+      })
 
-      expect(res.status).toEqual(200)
+      expect(res.status).toEqual(201)
```
Comment (saevarma) [inline apps/services/auth/ids-api/src/app/delegations/test/delegations-filters.spec.ts:174]: A RPC should respond with 200 OK, NestJS by default responds 201 for POST but you can either use our `@island.is/nest/swagger` `Documentation` decorator to set `response.status` as `200` or NestJS `HttpCode` decorator.

File: libs/auth-api-lib/src/lib/delegations/delegation-scope.service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (saevarma) [reply libs/auth-api-lib/src/lib/delegations/delegation-scope.service.ts]: My suggestion is not removing from the index if there was an error, as it breaks and goes to the catch block logs and returns an empty array. I feel like its cleaner and easier to read do the if check and indexing inside the try.

File: apps/services/auth/ids-api/src/app/delegations/delegation-verification.dto.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+import { AuthDelegationType } from '@island.is/shared/types'
+
+export class DelegationVerification {
+  @IsString()
+  @ApiProperty()
+  fromNationalId!: string
+
+  @IsArray()
+  @IsEnum(AuthDelegationType, { each: true })
+  @ApiProperty({ type: [AuthDelegationType] })
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply apps/services/auth/ids-api/src/app/delegations/delegation-verification.dto.ts]: Fixed?

**Review by saevarma — APPROVED** [review]
🚀

File: apps/services/auth/admin-api/infra/auth-admin-api.ts [hunk 9]
```diff
@@ -59,6 +59,12 @@ export const serviceSetup = (): ServiceBuilder<'services-auth-admin-api'> => {
         prod: 'IS/GOV/5402696029/Skatturinn/ft-v1',
       },
       COMPANY_REGISTRY_REDIS_NODES: REDIS_NODE_CONFIG,
+      SYSLUMENN_HOST: {
+        dev: 'https://api.syslumenn.is/staging',
```
Comment (brynjarorng) [inline apps/services/auth/admin-api/infra/auth-admin-api.ts:63]: Is this supposed to link to staging on dev?

File: apps/services/auth/admin-api/infra/auth-admin-api.ts [hunk 10]
```diff
@@ -59,6 +59,12 @@ export const serviceSetup = (): ServiceBuilder<'services-auth-admin-api'> => {
...
+        staging: 'https://api.syslumenn.is/staging',
+        prod: 'https://api.syslumenn.is',
+      },
+      SYSLUMENN_TIMEOUT: '10000',
```
Comment (brynjarorng) [inline apps/services/auth/admin-api/infra/auth-admin-api.ts:67]: 10 second timeout is quite long, is it really necessary to have it set this high?

File: apps/services/auth/delegation-api/infra/delegation-api.ts [hunk 11]
```diff
@@ -54,12 +54,20 @@ export const serviceSetup = (services: {
         prod: 'IS/GOV/5402696029/Skatturinn/ft-v1',
       },
       COMPANY_REGISTRY_REDIS_NODES: REDIS_NODE_CONFIG,
+      SYSLUMENN_HOST: {
+        dev: 'https://api.syslumenn.is/staging',
+        staging: 'https://api.syslumenn.is/staging',
+        prod: 'https://api.syslumenn.is',
+      },
+      SYSLUMENN_TIMEOUT: '10000',
```
Comment (brynjarorng) [inline apps/services/auth/delegation-api/infra/delegation-api.ts:62]: Staging referenced on dev again and timeout is quite high

File: apps/services/auth/personal-representative/infra/personal-representative.ts [hunk 12]
```diff
@@ -42,10 +42,18 @@ export const serviceSetup =
           prod: 'IS/GOV/5402696029/Skatturinn/ft-v1',
         },
         COMPANY_REGISTRY_REDIS_NODES: REDIS_NODE_CONFIG,
+        SYSLUMENN_HOST: {
+          dev: 'https://api.syslumenn.is/staging',
+          staging: 'https://api.syslumenn.is/staging',
+          prod: 'https://api.syslumenn.is',
+        },
+        SYSLUMENN_TIMEOUT: '10000',
```
Comment (brynjarorng) [inline apps/services/auth/personal-representative/infra/personal-representative.ts:50]: Same comment regarding staging and timeout

File: apps/services/auth/public-api/infra/auth-public-api.ts [hunk 13]
```diff
@@ -64,12 +63,20 @@ export const serviceSetup = (): ServiceBuilder<'services-auth-public-api'> => {
         // Origin for Android prod app
         'android:apk-key-hash:EsLTUu5kaY7XPmMl2f7nbq4amu-PNzdYu3FecNf90wU',
       ]),
+      SYSLUMENN_HOST: {
+        dev: 'https://api.syslumenn.is/staging',
+        staging: 'https://api.syslumenn.is/staging',
+        prod: 'https://api.syslumenn.is',
+      },
+      SYSLUMENN_TIMEOUT: '10000',
```
Comment (brynjarorng) [inline apps/services/auth/public-api/infra/auth-public-api.ts:71]: same again

**Review by kksteini — APPROVED** [review]
Codeowner files LGTM

File: apps/services/auth/admin-api/infra/auth-admin-api.ts (hunk 10 above)
Comment (valurefugl) [reply apps/services/auth/admin-api/infra/auth-admin-api.ts]: Good point, decreased to 3 s.

File: apps/services/auth/delegation-api/infra/delegation-api.ts (hunk 11 above)
Comment (valurefugl) [reply apps/services/auth/delegation-api/infra/delegation-api.ts]: Fixed.

File: apps/services/auth/public-api/infra/auth-public-api.ts (hunk 13 above)
Comment (valurefugl) [reply apps/services/auth/public-api/infra/auth-public-api.ts]: Fixed.

File: apps/services/auth/personal-representative/infra/personal-representative.ts (hunk 12 above)
Comment (valurefugl) [reply apps/services/auth/personal-representative/infra/personal-representative.ts]: Fixed.

**Review by brynjarorng — APPROVED** [review]
LGTM

**Pull request merged**